### PR TITLE
(CDAP-15127) Make remote runtime startup asynchronous

### PR DESCRIPTION
**Pull request by chtyim**

### CODE REVIEW

**Comment by chtyim** [comment]
Build https://builds.cask.co/browse/CDAP-DUT6953

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/remote/RemoteExecutionTwillRunnerService.java [hunk 1]
```diff
@@ -514,6 +478,124 @@ private static Location getKeysDirLocation(ProgramOptions programOptions, Locati
...
+              deleteRunningState(programRunId);
+              programStateWriter.error(programRunId, throwable);
+            }
+          });
+        } else {
+          // Otherwise, complete the startup task immediately
+          startupTaskCompletion.complete(null);
+        }
+
+        return createController(startupTaskCompletion);
```
Comment (poornachandra) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/remote/RemoteExecutionTwillRunnerService.java]: Looks like this controller should be added to the `controllers` map.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/remote/RemoteExecutionTwillRunnerService.java [hunk 2]
```diff
@@ -514,6 +478,124 @@ private static Location getKeysDirLocation(ProgramOptions programOptions, Locati
...
+    ControllerFactory(ProgramRunId programRunId, ProgramOptions programOptions, ClusterKeyInfo clusterKeyInfo) {
+      this.programRunId = programRunId;
+      this.programOptions = programOptions;
+      this.clusterKeyInfo = clusterKeyInfo;
+    }
+
+    @Override
+    public RemoteExecutionTwillController create(@Nullable Callable<Void> startupTask,
+                                                 long timeout, TimeUnit timeoutUnit) {
+      controllersLock.lock();
```
Comment (poornachandra) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/remote/RemoteExecutionTwillRunnerService.java]: `controllers` map is accessed without the lock in other places, why is this lock needed? Would be good to add a comment on what it is guarding.

**Review by poornachandra — COMMENTED** [review]
@chtyim Couple of comments

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/remote/RemoteExecutionTwillRunnerService.java (hunk 1 above)
Comment (chtyim) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/remote/RemoteExecutionTwillRunnerService.java]: It is added from the `createController` method.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/remote/RemoteExecutionTwillRunnerService.java (hunk 2 above)
Comment (chtyim) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/remote/RemoteExecutionTwillRunnerService.java]: The map is a concurrent map. The lock is only to make sure there is no multiple controller being created.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/remote/RemoteExecutionTwillRunnerService.java (hunk 1 above)
Comment (chtyim) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/remote/RemoteExecutionTwillRunnerService.java]: Oh, wait.. it is missing.

**Review by poornachandra — APPROVED** [review]

**Pull request merged**